### PR TITLE
add proper project preferences support

### LINKED ISSUE #125: Fix misalign of what attribute id == project preferences should do vs Student.preferences
Opey forgot he added a Student.preferences and built the MockStudentProvider assuming project preferences would be treated just like any other attribute - this ticket is to pick a lane and make it true

### PR DESCRIPTION
**Pull request by Opeyem1a**
closes #125 

tldr; I got greedy and earlier thought I could represent project preferences as just any other attribute - we can't. It's important for many algorithms that we know what the project preferences are.

This PR conclusively makes `Student.project_preferences` *the* way to represent this info

Also included:
- MockStudentProvider can now mock students having preferences (it's more obvious how to do this, too, RE: #101 )
- Adding validations to mock student provider settings
- Adding compatibility validation between initial teams provider and student provider
- Minor improvements to validation utils

### CODE REVIEW

**Comment by Opeyem1a** [comment]
Current dependencies on/for this PR:
* main
  * **PR #157** <a href="https://app.graphite.dev/github/pr/Teamable-Analytics/algorithms/157" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Teamable-Analytics/algorithms/157?utm_source=stack-comment).

**Review by justino599 — APPROVED** [review]
LGTM
